sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnpredictableHashSaltCheck.kt

```diff
@@ -0,0 +1,150 @@
+            return
+        }
+
+        val bindingContext = kotlinFileContext.bindingContext
+        val saltArg = resolvedCall.valueArgumentsByIndex?.elementAtOrNull(saltArgIndex)
+            ?.arguments?.elementAtOrNull(0)?.getArgumentExpression() ?: return
+
+        val predictedSaltValue = saltArg.predictRuntimeValueExpression(bindingContext) ?: return
+
+        if (predictedSaltValue.isInitializedFromString(bindingContext)) {
+            kotlinFileContext.reportIssue(
+                saltArg,
+                MSG_MAKE_UNPREDICTABLE,
+                listOf(kotlinFileContext.secondaryOf(predictedSaltValue, SMSG_PREDICTABLE_SALT))
+            )
+            return
+        }
+
+        val saltInitializer = predictedSaltValue.getResolvedCall(bindingContext)
+            ?.takeIf { it matches BYTE_ARRAY_CONSTRUCTOR } ?: return
+
+        if (saltInitializer.byteArrayInitSizeTooSmall(bindingContext)) {
+            kotlinFileContext.reportIssue(saltArg, MSG_MIN_LEN, listOf(kotlinFileContext.secondaryOf(predictedSaltValue)))
+        }
+
+        if (saltInitializer matches BYTE_ARRAY_CONSTRUCTOR_SIZE_ARG_ONLY && saltArg.isInitializedPredictably(bindingContext)) {
+            kotlinFileContext.reportIssue(
+                saltArg,
+                MSG_MAKE_UNPREDICTABLE,
+                listOf(kotlinFileContext.secondaryOf(predictedSaltValue, SMSG_PREDICTABLE_SALT))
+            )
+        }
+    }
+
+    /**
+     * Checks whether the expression is a call, matches the FunMatchers in [STRING_TO_BYTE_FUNS] and is called on a constant string value.
+     */
+    private fun KtExpression.isInitializedFromString(bindingContext: BindingContext) =
+        getCalleeOrUnwrappedGetMethod(bindingContext)?.let { callee ->
+            STRING_TO_BYTE_FUNS.any { it.matches(callee) } &&
+                (getCall(bindingContext)?.explicitReceiver as? ExpressionReceiver)?.expression?.predictRuntimeStringValue(bindingContext) != null
+        } ?: false
+
+    /**
+     * Checks whether the first argument of the call is an integer that is at least 16
+     */
+    private fun ResolvedCall<*>.byteArrayInitSizeTooSmall(bindingContext: BindingContext) =
+        (this.valueArgumentsByIndex
+            ?.firstOrNull()
+            ?.arguments?.firstOrNull()
+            ?.getArgumentExpression()
+            ?.predictRuntimeIntValue(bindingContext) ?: 16) < 16
+
+    /**
+     * Checks whether the variable has been initialized with the help of a secure random function
+     */
+    private fun KtExpression.isInitializedPredictably(bindingContext: BindingContext): Boolean {
+        if (this is KtNameReferenceExpression) {
+            val usages =
+                this.getParentOfType<KtBlockExpression>(false)
+                    ?.collectDescendantsOfType<KtNameReferenceExpression> { it.getReferencedName() == this.getReferencedName() }
+                    ?: return false
+
+            for (usage in usages) {
```

Comment:
It's minor, but it looks like an ordinary for cycle breaks the style. Could this be a 

```
if (usages.filter { this !== it }
            .any { it.getParentOfType<KtCallExpression>(false).getResolvedCall(bindingContext) matches SECURE_RANDOM_FUNS }) {
                return false
            }
```

Reply:
Yea not sure what went through my head there

Reply:
Ah actually I can tell you why I did it like this. Say I have:
```
    val saltInitializedTooLate = ByteArray(16)
    PBEParameterSpec(saltInitializedTooLate, 10000) // Noncompliant {{Make this salt unpredictable.}}
    SecureRandom().nextBytes(saltInitializedTooLate)
```
Here I still want an issue to be raised. To make sure I don't accidentally think that the salt was initialized, I want to stop analyzing the references as soon as I hit the usage in the constructor (line 2). That way I never look at line 3. If I only check if the list contains any instance of the array's initialization, I'd find that yes, that is the case - which is not what I want to know. In fact I have added this exact test case to the PR